### PR TITLE
qdel optimize

### PR DESCRIPTION
**Pull request by FeenieRU**
## About The Pull Request

That PR optimizes qdel and SSgarbage procs.

Based on:
https://github.com/tgstation/tgstation/pull/79568
https://github.com/tgstation/tgstation/pull/76956
https://github.com/tgstation/tgstation/pull/80443
https://github.com/tgstation/tgstation/pull/80628

## Why It's Good For The Game

Better performance. Tested on downstream: https://github.com/CeladonSS13/Shiptest/pull/1025

## Changelog

:cl:
code: Changing qdel() and SSgarbage procs
code: rewrite /Destroy(force, silent) to /Destroy(force)
/:cl:

### CODE REVIEW

**Comment by FeenieRU** [comment]
Main problem for compiling - no **libdreamluau.so** per compiling

`tools/ci/install_dreamluau.sh`
`tools/tgs_scripts/PreCompile.sh`

**Comment by thgvr** [comment]
A better solution than adding another auxtool modification maintained by one person we don't even have contact with would be actually just fixing harddels

File: code/controllers/subsystem/garbage.dm [hunk 1]
```diff
@@ -324,79 +345,86 @@ SUBSYSTEM_DEF(garbage)
...
-
 /// Should be treated as a replacement for the 'del' keyword.
 ///
 /// Datums passed to this will be given a chance to clean up references to allow the GC to collect them.
-/proc/qdel(datum/D, force=FALSE, ...)
-	if(!istype(D))
-		del(D)
+/proc/qdel(datum/to_delete, force = FALSE)
+	if(!istype(to_delete))
+		DREAMLUAU_CLEAR_REF_USERDATA(to_delete)
```
Comment (MarkSuckerberg) [inline code/controllers/subsystem/garbage.dm]: This is the only time Dreamluau is ever called in the code outside of the bindings file. Again, this *only* lets Dreamluau check if it has any references to the object in Lua code, nothing else.

To remove Dreamluau so that this PR stands on its own, just get rid of this single line of code and the _dreamluau.dm file, along with the scripts in tools.

**Review by MarkSuckerberg — CHANGES_REQUESTED** [review]
Alright, so: this PR is alright, but you misunderstand what you ported here. Dreamluau doesn't change garbage collection/qdel at all, except for cleaning up its own references in the luau code. All the optimizations besides it that are in this PR is fine, but Dreamluau 100% does not need to be included here, there is absolutely no benefit it gives us, in fact, the only thing it could possibly do is slow down GC, even if by a trivial amount, because again, it *only* cleans up its own references, it doesn't touch the DM code at all.

Basically, this PR is in dire need of atomization. You can port Dreamluau in a different PR if you'd like, but this PR will not be merged with both qdel optimizations and a full port of an entirely unused external library.

Also, please make sure to credit the original TG PRs that added these changes, if you could please.

**Comment by FeenieRU** [comment]
Done. Sorry, I'm low-sorted coder.

**Review by MarkSuckerberg — APPROVED** [review]
Understandable, and it's looking a lot better. Don't forget `code/__HELPERS/auxtools.dm` before it's merged, and I'll see if we can get this testmerged soon.

**Comment by FeenieRU** [comment]
> Don't forget `code/__HELPERS/auxtools.dm`

Done

**Comment by FeenieRU** [comment]
изображение

Whyyyy?! How much I should update it?

**Comment by Erikafox** [comment]
image
I assume the create and destroy test is failing on the same spot each time

**Comment by FeenieRU** [comment]
Any ideas for fix?

**Comment by FalloutFalcon** [comment]
I don't see any unit tests ran on there and its a drafted pr it really doesn't seem like its been "tested" there.
skimming the code I don't see anything that's clearly more performant though it looks like it should return earlier in some situations. What are you basing that off of?
> Better performance. Tested on downstream: https://github.com/CeladonSS13/Shiptest/pull/1025

This also ports parts of 
https://github.com/tgstation/tgstation/pull/80443
https://github.com/tgstation/tgstation/pull/76956
https://github.com/tgstation/tgstation/pull/80628
but not everything which is my initial guess to why copy pasting the code didn't work

**Comment by FeenieRU** [comment]
> I don't see any unit tests ran on there and its a drafted pr it really doesn't seem like its been "tested" there. skimming the code I don't see anything that's clearly more performant though it looks like it should return earlier in some situations. What are you basing that off of?
> 
> > Better performance. Tested on downstream: [CeladonSS13#1025](https://github.com/CeladonSS13/Shiptest/pull/1025)
> 
> This also ports parts of [tgstation/tgstation#80443](https://github.com/tgstation/tgstation/pull/80443) [tgstation/tgstation#76956](https://github.com/tgstation/tgstation/pull/76956) [tgstation/tgstation#80628](https://github.com/tgstation/tgstation/pull/80628) but not everything which is my initial guess to why copy pasting the code didn't work

Done

**Review by FalloutFalcon — APPROVED** [review]
Looks good I will give it a test merge today or this weekend, only small thing is the existing harddel but ill fix it if you dont.

**Comment by FalloutFalcon** [comment]
and and update your change log

**Comment by FeenieRU** [comment]
> and and update your change log

ready

**Comment by FeenieRU** [comment]
> Looks good I will give it a test merge today or this weekend, only small thing is the existing harddel but ill fix it if you dont.

honestly, idk why it happens

**Comment by MarkSuckerberg** [comment]
I fixed the harddel in #3572, this is fine to TM, ideally with my PR though

**Comment by FeenieRU** [comment]
everything is okay?

**Comment by FeenieRU** [comment]
изображение

are you kidding me...

**Comment by FeenieRU** [comment]
FINALLY

**Comment by FeenieRU** [comment]
what's up with docker?..

**Comment by Sun-Soaked** [comment]
> what's up with docker?..

I made a mistake clearing up mine references, give me a moment to post an emergency pr
